ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/CommandTopicBackup.java

```diff
@@ -0,0 +1,210 @@
+        LOG.info("New backup file created: {}", replayFile.getPath());
+      }
+    } else if (latestReplayIdx > 0) {
+      // clear latest replay from memory
+      latestReplay.clear();
+      latestReplayIdx = 0;
+    }
+
+    try {
+      replayFile.write(record.key(), record.value());
+    } catch (final IOException e) {
+      LOG.warn("Failed to write to file {}. The command topic backup is not complete. "
+              + "Make sure the file exists and has permissions to write. KSQL must be restarted "
+              + "afterwards to complete the backup process. Error = {}",
+          replayFile.getPath(), e.getMessage());
+    }
+  }
+
+  private void createNewBackupFile() {
+    try {
+      replayFile.close();
+    } catch (IOException e) {
+      LOG.warn("Couldn't close the current backup file {}. Error = {}",
+          replayFile.getPath(), e.getMessage());
+    }
+
+    replayFile = newReplayFile();
+
+    if (latestReplay.size() > 0 && latestReplayIdx > 0) {
+      try {
+        replayFile.write(latestReplay.subList(0, latestReplayIdx));
+      } catch (final IOException e) {
+        LOG.warn("Couldn't write the latest replayed commands to the new backup file {}. "
+                + "Make sure the file exists and has permissions to write. "
+                + "KSQL must be restarted afterwards to complete the backup process. Error = {}",
+            replayFile.getPath(), e.getMessage());
+      }
+    }
+
+    // clear latest replay from memory
+    latestReplay.clear();
+    latestReplayIdx = 0;
+  }
+
+  @VisibleForTesting
+  BackupReplayFile openOrCreateReplayFile() {
+    final Optional<BackupReplayFile> latestFile = latestReplayFile();
+    if (latestFile.isPresent()) {
+      return latestFile.get();
+    }
+
+    return newReplayFile();
+  }
+
+  private BackupReplayFile newReplayFile() {
+    return new BackupReplayFile(Paths.get(
+        backupLocation.getAbsolutePath(),
+        String.format("%s%s_%s", PREFIX, topicName, ticker.read())
+    ).toFile());
+  }
+
+  private Optional<BackupReplayFile> latestReplayFile() {
+    final String prefixFilename = String.format("%s%s_", PREFIX, topicName);
+    final File[] files = backupLocation.listFiles(
+        (f, name) -> name.toLowerCase().startsWith(prefixFilename));
+
+    File latestBakFile = null;
+    if (files != null) {
+      long latestTs = 0;
+      for (int i = 0; i < files.length; i++) {
+        final File bakFile = files[i];
+        final String bakTimestamp = bakFile.getName().substring(prefixFilename.length());
+
+        try {
+          final Long ts = Long.valueOf(bakTimestamp);
+          if (ts > latestTs) {
+            latestTs = ts;
+            latestBakFile = bakFile;
+          }
+        } catch (final NumberFormatException e) {
+          LOG.warn(
+              "Invalid timestamp '{}' found in backup replay file (file ignored): {}",
+              bakTimestamp, bakFile.getName());
+          continue;
+        }
+      }
+    }
```

Comment:
another code style thing which might make this a little easier to read:
```java
File latestFile = Arrays.stream(files).max(Comparator.comparing(CommandTopicBackup::extractTimestamp));
```

Reply:
Looks nice!. But, how do I ignore a `extractTimestamp` that failed? Say file is `file_123xx`. I would need to extract the timestamp as String, then a filter that checks isNumeric, then a map to convert to Long, then a max(). Is there a better way to ignore it like your code?

Reply:
good point... I'm not long certain this is any better than what you have, but one option is to make `extractTimestamp` return `Long.MIN_VALUE` if there's a failure and ensuring the one we got at the end is valid. Feel free to just close this comment :)